library/network/src/lib/y2firewall/firewalld/zone_reader.rb

```diff
@@ -0,0 +1,132 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2017 SUSE LLC
+#
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact SUSE.
+#
+# To contact SUSE about this file by physical or electronic mail, you may find
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "y2firewall/firewalld/api"
+require "y2firewall/firewalld/zone"
+
+module Y2Firewall
+  class Firewalld
+    # Class to help parsing firewall-cmd --list-all-zones output
+    class ZoneReader
+      include Yast::Logger
+
+      # @return [Array<String] configured zone names
+      attr_accessor :zone_names
+      # @return [Hash<String,Hash>] stores the parsed configuration for eac
+      #   zone indexed by its name
+      attr_accessor :zone_entries
+      # @return [String] zones definition to be parsed for initializing the
+      #   zone objects
+      attr_accessor :zones_definition
+
+      BOOLEAN_ATTRIBUTES = ["icmp-block-inversion", "masquerade"].freeze
+      MULTIPLE_ENTRIES = ["rich_rules", "forward_ports"].freeze
+
+      # Constructor
+      #
+      # @param zone_names [Array<String>] zone names
+      # @param zones_definition [String] text with the complete definition of
+      #   existing zones.
+      def initialize(zone_names, zones_definition)
+        @zone_names = zone_names
+        @zones_definition = zones_definition
+        @zone_entries = {}
+      end
+
+      # It reads the zone definition instantiating the defined zones and
+      # settings their attributes.
+      #
+      # @return [Array<Y2Firewall::Firewalld::Zone>]
+      def read
+        return [] if !@zone_names || @zone_names.empty?
+        parse_zones
+        initialize_zones
+      end
+
+    private
+
+      # Iterates over the zones definition filling the zone entries with the
+      # parsed information of each zone
+      def parse_zones
+        current_zone = nil
+        current_attribute = nil
+        zones_definition.each do |line|
+          next if line.lstrip.empty?
+          # If  the entry looks like a zone name
+          if line.start_with?(/\w/)
+            attribute, _value = line.split(/\s*\(active\)\s*$/)
+            attribute = nil unless zone_names.include?(attribute)
+            current_zone = attribute
```

Comment:
If the `attribute` value is not relevant anymore for this iteration (when the zone name is not included in `zone_names`), I would write something like `current_zone = zone_names.include?(attribute) ? attribute : nil`.